### PR TITLE
fix a-doc listing plugins and add tests

### LINKED ISSUE #69031: "ansible-doc --json aws_s3_bucket_info" crashes
##### SUMMARY
When running `ansible-doc --json aws_s3_bucket_info` in stable-2.9:
```
ansible-doc 2.9.7.post0
  config file = None
  configured module search path = ['/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/lib/ansible
  executable location = /ansible/bin/ansible-doc
  python version = 3.8.2 (default, Apr  8 2020, 14:31:25) [GCC 9.3.0]
No config file found; using defaults
/ansible/lib/ansible/cli/doc.py:219: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  doc_data['return'] = yaml.load(doc_data['return'])
ERROR! Unexpected Exception, this is probably a bug: Object of type datetime is not JSON serializable
the full traceback was:

Traceback (most recent call last):
  File "/ansible/bin/ansible-doc", line 123, in <module>
    exit_code = cli.run()
  File "/ansible/lib/ansible/cli/doc.py", line 223, in run
    jdump(plugin_docs)
  File "/ansible/lib/ansible/cli/doc.py", line 36, in jdump
    display.display(json.dumps(text, sort_keys=True, indent=4))
  File "/usr/lib/python3.8/json/__init__.py", line 234, in dumps
    return cls(
  File "/usr/lib/python3.8/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/usr/lib/python3.8/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/usr/lib/python3.8/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/lib/python3.8/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/lib/python3.8/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  [Previous line repeated 1 more time]
  File "/usr/lib/python3.8/json/encoder.py", line 325, in _iterencode_list
    yield from chunks
  File "/usr/lib/python3.8/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/lib/python3.8/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/usr/lib/python3.8/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type datetime is not JSON serializable
```

I guess the problem is that the docs YAML contains something that is interpreted as a datetime by the YAML parser.

IMO ansible-doc should not crash, but instead convert everything that is not JSON-serializable to a string (or omit it).

This is important for everything that uses `ansible-doc --json` to extract docs (AFAIK Galaxy plans to do that).

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
ansible-doc

##### ANSIBLE VERSION
```paste below
2.9
(probably also devel)
```

### PR DESCRIPTION
**Pull request by bcoca**
and  fix non safe_load yaml call

also fixes #69031, alternate to #69032

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

ansible-doc

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure due to traceback in the `ansible-doc` integration test:

```
ERROR! Unexpected Exception, this is probably a bug: Can't mix strings and bytes in path components
```

File: lib/ansible/cli/doc.py [hunk 1]
```diff
@@ -300,9 +300,9 @@ def get_plugin_metadata(plugin_type, plugin_name):
 
     @staticmethod
     def namespace_from_plugin_filepath(filepath, plugin_name, basedir):
-        if not basedir.endswith('/'):
-            basedir += '/'
-        rel_path = filepath.replace(basedir, '')
+        if not basedir.endswith(b'/'):
```
Comment (jborean93) [inline lib/ansible/cli/doc.py]: Don't we usually subscribe to a "unicode sandwich" strategy where functions pass around unicode strings and the function converts to bytes or native strings when it crosses an external boundary? I've also seen the `b_` or `n_` prefix being used if a function specifically accepts a byte or native string but not sure if that's an actual guideline we have going forward.

**Comment by mattclay** [comment]
@bcoca Can the modules/plugins used for the tests be trimmed down? It looks like they contain code not needed for the test.

**Comment by bcoca** [comment]
@mattclay i'll try again, i kept getting errors from the sanity tests so i kept them 'semi functional'

File: lib/ansible/collections/__init__.py [hunk 2]
```diff
@@ -6,8 +6,9 @@
 
 import os
 
+from ansible.module_utils._text import to_bytes
 
-FLAG_FILES = frozenset(['MANIFEST.json', 'galaxy.yml'])
+FLAG_FILES = frozenset([b'MANIFEST.json', b'galaxy.yml'])
```
Comment (mattclay) [inline lib/ansible/collections/__init__.py]:
```suggestion
B_FLAG_FILES = frozenset([b'MANIFEST.json', b'galaxy.yml'])
```

File: lib/ansible/cli/doc.py [hunk 3]
```diff
@@ -18,33 +18,36 @@
...
 
 def jdump(text):
-    display.display(json.dumps(text, sort_keys=True, indent=4))
+    display.display(json.dumps(text, cls=AnsibleJSONEncoder, sort_keys=True, indent=4))
 
 
 def add_collection_plugins(plugin_list, plugin_type, coll_filter=None):
 
     # TODO: take into account routing.yml once implemented
     colldirs = list_collection_dirs(coll_filter=coll_filter)
```
Comment (mattclay) [inline lib/ansible/cli/doc.py]:
```suggestion
    b_colldirs = list_collection_dirs(coll_filter=coll_filter)
```

File: lib/ansible/cli/doc.py [hunk 4]
```diff
@@ -18,33 +18,36 @@
...
-    for path in colldirs:
-        collname = get_collection_name_from_path(path)
+    for b_path in colldirs:
```
Comment (mattclay) [inline lib/ansible/cli/doc.py]:
```suggestion
    for b_path in b_colldirs:
```

File: lib/ansible/collections/__init__.py [hunk 5]
```diff
@@ -18,9 +19,10 @@ def is_collection_path(path):
     """
 
     is_coll = False
-    if os.path.isdir(path):
+    b_path = to_bytes(path)
+    if os.path.isdir(b_path):
         for flag in FLAG_FILES:
```
Comment (mattclay) [inline lib/ansible/collections/__init__.py]:
```suggestion
        for b_flag in B_FLAG_FILES:
```

File: lib/ansible/collections/__init__.py [hunk 6]
```diff
@@ -18,9 +19,10 @@ def is_collection_path(path):
...
-            if os.path.exists(os.path.join(path, flag)):
+            if os.path.exists(os.path.join(b_path, flag)):
```
Comment (mattclay) [inline lib/ansible/collections/__init__.py]:
```suggestion
            if os.path.exists(os.path.join(b_path, b_flag)):
```

File: lib/ansible/utils/collection_loader.py [hunk 7]
```diff
@@ -550,9 +550,9 @@ def get_collection_name_from_path(path):
     :param n_path: native-string path to evaluate for collection containment
     :return: collection name or None
     """
-    n_collection_paths = [to_native(os.path.realpath(to_bytes(p))) for p in AnsibleCollectionLoader().n_collection_paths]
+    n_collection_paths = [to_native(os.path.abspath(to_bytes(p))) for p in AnsibleCollectionLoader().n_collection_paths]
```
Comment (mattclay) [inline lib/ansible/utils/collection_loader.py]: Why the change here and below from `realpath` to `abspath`?

**Review by mattclay — CHANGES_REQUESTED** [review]
Overall looks good. Only one question, and a few changes to variable names for consistency.

File: lib/ansible/utils/collection_loader.py (hunk 7 above)
Comment (bcoca) [reply lib/ansible/utils/collection_loader.py]: to allow for symlinks

Comment (mattclay) [reply lib/ansible/utils/collection_loader.py]: Is that related to the ansible-doc fixes? It seems like a separate issue.

Comment (bcoca) [reply lib/ansible/utils/collection_loader.py]: related to testing them, i  had complaint docs didn't work for several people, while loading the collection plugins did work, i traced it back to them using symlinks.

Comment (mattclay) [reply lib/ansible/utils/collection_loader.py]: Could you put those fixes in a separate PR and include tests with it to prevent regressions?

**Review by mattclay — APPROVED** [review]
